internal/cloud/realm/hosting.go

```diff
@@ -0,0 +1,213 @@
+package realm
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/10gen/realm-cli/internal/utils/api"
+)
+
+const (
+	hostingPathPattern       = appPathPattern + "/hosting"
+	hostingAssetsPathPattern = hostingPathPattern + "/assets"
+	hostingAssetPathPattern  = hostingAssetsPathPattern + "/asset"
+	hostingCachePathPattern  = hostingPathPattern + "/cache"
+
+	hostingQueryRecursive = "recursive"
+
+	paramMetadata = "meta"
+	paramPath     = "path"
+)
+
+// HostingAssetData is a Realm app's hosting asset data
+type HostingAssetData struct {
+	FilePath     string `json:"path"`
+	FileHash     string `json:"hash,omitempty"`
+	FileSize     int64  `json:"size,omitempty"`
+	LastModified int64  `json:"last_modified,omitempty"`
+}
+
+// HostingAsset is a Realm app's hosting asset
+type HostingAsset struct {
+	HostingAssetData
+	AppID string                 `json:"appId,omitempty"`
+	Attrs HostingAssetAttributes `json:"attrs"`
+	URL   string                 `json:"url,omitempty"`
+}
+
+// HostingAssetAttribute is a Realm app's hosting asset attribute
+type HostingAssetAttribute struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+func (c *client) HostingAssets(groupID, appID string) ([]HostingAsset, error) {
+	res, err := c.do(
+		http.MethodGet,
+		fmt.Sprintf(hostingAssetsPathPattern, groupID, appID),
+		api.RequestOptions{Query: map[string]string{hostingQueryRecursive: trueVal}},
+	)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != http.StatusOK {
+		return nil, api.ErrUnexpectedStatusCode{"get hosting assets", res.StatusCode}
+	}
+	defer res.Body.Close()
+
+	var assets []HostingAsset
+	if err := json.NewDecoder(res.Body).Decode(&assets); err != nil {
+		return nil, err
+	}
+	return assets, nil
+}
+
+func (c *client) HostingAssetUpload(groupID, appID, rootDir string, asset HostingAsset) error {
+	file, err := os.Open(filepath.Join(rootDir, asset.FilePath))
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	data, err := json.Marshal(HostingAsset{
+		AppID: appID,
+		HostingAssetData: HostingAssetData{
+			FilePath: asset.FilePath,
+			FileHash: asset.FileHash,
+			FileSize: asset.FileSize,
+		},
+		Attrs: asset.Attrs,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Construct a pipe stream: the reader side will be consumed and sent as the body of the outgoing request,
+	// and the writer side we can use to asynchronously populate it.
+	pipeReader, pipeWriter := io.Pipe()
+
+	bodyWriter := multipart.NewWriter(pipeWriter)
+	errChan := make(chan error, 2)
+
+	go func() {
+		defer func() {
+			bodyWriter.Close()
+			pipeWriter.Close()
+		}()
+		// Create the first part and write the metadata into it
+		mw, err := bodyWriter.CreateFormField(paramMetadata)
+		if err != nil {
+			errChan <- fmt.Errorf("failed to create metadata multipart field: %w", err)
```

Comment:
should we break on these errors as well?

Reply:
i was thinking that as well, but based on the current implementation it appears we don't do that, and i was worried about changing any existing behavior too much there.  it seemed almost intentional that we try to process all the files first, and just alert/warn on any errors